Wrap FrozenError, to state that the container itself is frozen

cllns opened the pull request:
Before:
```
FrozenError: 
  can't modify frozen Concurrent::Hash: {"foo"=>#<Dry::Container::Item::Callable:0x00007f9a9f2eaea0 @item="bar", @options={:call=>false}>}
```

After:
```
FrozenError:
  can't modify frozen Dry::Container (when attempting to register 'baz')
```

That a `Dry::Container` is implemented with `Concurrent::Hash` is an implemention detail. I think wrapping it helps users of the library see that this is an expected exception, rather than an error within the library. And the text in the parentheses should help users when they encounter this error.

I also split up the specs and removed a conditional for whether Ruby is 2.5+, since this gem only supports 2.5+ now.

(Relevant commit where `.freeze` was implemented: https://github.com/dry-rb/dry-container/commit/6fc5cd2650b1888ac6f4fc56592878695d296979)

cllns committed:
Wrap FrozenError, to state that the container itself is frozen

Also include which key was attempted to be registered, to assist with
debugging

cllns commented:
Another fun enhancement to this would be to store a `@_freeze_caller = caller` when freezing it, then surface it when this error occurs, telling users where they froze the container.

solnic commented on lib/dry/container/mixin.rb:
```diff
@@ -105,6 +105,8 @@ def register(key, contents = nil, options = EMPTY_HASH, &block)
 
         self
+      rescue FrozenError
+        raise FrozenError, "can't modify frozen Dry::Container (when attempting to register '#{key}')"
```
```suggestion
        raise FrozenError, "can't modify frozen #{self.class} (when attempting to register '#{key}')"
```

cllns replied:
I actually tried this originally but it's very tricky to test with an anonymous (unnamed) class. 

Changing _just_ this line causes three test failures, which are simple enough to work around individually, but hard to work around in a way that works for all of them (since this is a shared example, it has to work for all of them).

1. Mixin extended: `FrozenError: can't modify frozen Class (when attempting to register 'baz')`
2. Mixin included `FrozenError: can't modify frozen #<Class:0x00007fc6af9f5ec0> (when attempting to register 'baz')`
3. Class `FrozenError: can't modify frozen Dry::Container (when attempting to register 'baz')`

I tried redefining the `klass` for this test but that won't work because the klass is contructed in different ways for each. And defining the `klass` with `class MyTestClass; include Dry::Container::Mixin; end;` doesn't work because then we're redefining the constant, and, yanno, we freeze the klass in these tests. We could `Object.remove_const` but I think we want go down that path for such a small feature.

Went with a simpler, though less robust method, of using a regex to match the message is right _except_ for the actual class name, since there are so many valid options there. I went for `\S+` (one or more non-whitespace characters). I think that should work? Don't think `self.class` will ever return something with a space in it, but I could be wrong. Can change that to some other regex metacharacter.

solnic reviewed (commented):
Thanks this is a good improvement! Left one suggestion.

cllns committed:
Use self.class instead of 'Dry::Container' in error

This is more accurate, since dry-container works as a:
- class,
- module include,
- module extend

cllns commented:
Rebased on master to resolve conflicts
